S3/S3.py

```diff
@@ -870,43 +874,7 @@ def urlencode_string(self, string, urlencoding_mode = None):
             ## Don't do any pre-processing
             return string
 
-        encoded = ""
-        ## List of characters that must be escaped for S3
-        ## Haven't found this in any official docs
-        ## but my tests show it's more less correct.
-        ## If you start getting InvalidSignature errors
-        ## from S3 check the error headers returned
-        ## from S3 to see whether the list hasn't
-        ## changed.
-        for c in string:    # I'm not sure how to know in what encoding
-                    # 'object' is. Apparently "type(object)==str"
-                    # but the contents is a string of unicode
-                    # bytes, e.g. '\xc4\x8d\xc5\xafr\xc3\xa1k'
-                    # Don't know what it will do on non-utf8
-                    # systems.
-                    #           [hope that sounds reassuring ;-)]
-            o = ord(c)
-            if (o < 0x20 or o == 0x7f):
-                if urlencoding_mode == "fixbucket":
-                    encoded += "%%%02X" % o
-                else:
-                    error(u"Non-printable character 0x%02x in: %s" % (o, string))
-                    error(u"Please report it to s3tools-bugs@lists.sourceforge.net")
-                    encoded += replace_nonprintables(c)
-            elif (o == 0x20 or  # Space and below
-                o == 0x22 or    # "
-                o == 0x23 or    # #
-                o == 0x25 or    # % (escape character)
-                o == 0x26 or    # &
-                o == 0x2B or    # + (or it would become <space>)
-                o == 0x3C or    # <
-                o == 0x3E or    # >
-                o == 0x3F or    # ?
-                o == 0x60 or    # `
-                o >= 123):      # { and above, including >= 128 for UTF-8
-                encoded += "%%%02X" % o
-            else:
-                encoded += c
+        encoded = quote_plus(string, safe="~/")
```

Comment:
oh interesting.  I wonder if this is the source of all the InvalidSignature errors we're seeing recently, that look perfectly valid to me?